pino-roll.js

```diff
@@ -0,0 +1,90 @@
+'use strict'
+
+const SonicBoom = require('sonic-boom')
+const { buildFileName, detectLastNumber, parseSize, parseFrequency } = require('./lib/utils')
+
+/**
+ * @typedef {object} Options
+ *
+ * @property {string} file - Absolute or relative path to the log file.
+ * Your application needs the write right on the parent folder.
+ * Number will be appened to this file name.
+ * When the parent folder already contains numbered files, numbering will continue based on the highest number.
+ * If this path does not exist, the logger with throw an error unless you set `mkdir` to `true`.
+ *
+ * @property {string|number} size? - When specified, the maximum size of a given log file.
+ * Can be combined with frequency.
+ * Use 'k', 'm' and 'g' to express values in KB, MB or GB.
+ * Numerical values will be considered as MB.
+ *
+ * @property {string|number} frequency? - When specified, the amount of time a given log file is used.
+ * Can be combined with size.
+ * Use 'daily' or 'hourly' to rotate file every day (or every hour).
+ * Existing file within the current day (or hour) will be re-used.
+ * Numerical values will be considered as a number of milliseconds.
+ * Using a numerical value will always create a new file upon startup.
+ *
+ * @property {string} extension? - When specified, appends a file extension after the file number.
+ */
+
+/**
+ * @typedef {Options & import('sonic-boom').SonicBoomOpts} PinoRollOptions
+ */
+
+/**
+ * Creates a Pino transport (a Sonic-boom stream) to writing into files.
+ * Automatically rolls your files based on a given frequency, size, or both.
+ *
+ * @param {PinoRollOptions} options - to configure file destionation, and rolling rules.
+ * @returns {SonicBoom} the Sonic boom steam, usabled as Pino transport.
+ */
+module.exports = async function ({ file, size, frequency, extension, ...opts } = {}) {
+  let frequencySpec = parseFrequency(frequency)
+
+  let number = await detectLastNumber(file, frequencySpec?.start)
+
+  let currentSize = 0
+  const maxSize = parseSize(size)
+
+  const destination = new SonicBoom({ ...opts, dest: buildFileName(file, number, extension) })
+
+  let rollTimeout
+  if (frequencySpec) {
+    destination.once('close', () => {
+      clearTimeout(rollTimeout)
+    })
+    scheduleRoll()
+  }
+
+  if (maxSize) {
+    // patching sonic-boom stream is fragile, but it's the most performant way to "watch" the file's size
+    const originalRelease = destination.release.bind(destination)
```

Comment:
I would really _not_ recommend anybody to do this. `destination.release` is one of the most tricky methods of sonic-boom. Can we add a feature there to simplify this?

Reply:
https://github.com/pinojs/sonic-boom/pull/140 is an attempt to make this information available without forcing use to monkey-patch SonicBoom.